lxd/openfga/client.go

```diff
@@ -0,0 +1,174 @@
+package openfga
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"net/url"
+	"strings"
+
+	openfga "github.com/openfga/go-sdk"
+	"github.com/openfga/go-sdk/credentials"
+
+	"github.com/canonical/lxd/shared/logger"
+)
+
+type Client struct {
+	apiClient   *openfga.APIClient
+	authModelID string
+}
+
+// NewClient returns a new Client struct containing an OpenFGA APIClient.
+func NewClient(apiURL string, apiToken string, storeID string) (*Client, error) {
+	u, err := url.Parse(apiURL)
+	if err != nil {
+		return nil, fmt.Errorf("Failed parsing URL: %w", err)
+	}
+
+	conf := openfga.Configuration{
+		ApiScheme: u.Scheme,
+		ApiHost:   u.Host,
+		StoreId:   storeID,
+	}
+
+	if apiToken != "" {
+		conf.Credentials = &credentials.Credentials{
+			Method: credentials.CredentialsMethodApiToken,
+			Config: &credentials.Config{
+				ApiToken: apiToken,
+			},
+		}
+	}
+
+	config, err := openfga.NewConfiguration(conf)
+	if err != nil {
+		return nil, err
+	}
+
+	client := openfga.NewAPIClient(config)
+
+	// Write authorization model
+	var body openfga.WriteAuthorizationModelRequest
+
+	err = json.Unmarshal([]byte(authModel), &body)
+	if err != nil {
+		return nil, err
+	}
+
+	data, _, err := client.OpenFgaApi.WriteAuthorizationModel(context.Background()).Body(body).Execute()
+	if err != nil {
+		return nil, err
+	}
+
+	authModelID := data.GetAuthorizationModelId()
+
+	return &Client{apiClient: client, authModelID: authModelID}, nil
+}
+
+// HasPermission returns true if the user has the relation to the object.
+func (c *Client) HasPermission(user string, relation Relation, object string) bool {
+	fullUser := UserObject(user)
+
+	logger.Debug("Checking ReBAC permission", logger.Ctx{"user": fullUser, "relation": relation, "object": object})
+
+	body := openfga.CheckRequest{
+		AuthorizationModelId: openfga.PtrString(c.authModelID),
+		TupleKey: openfga.TupleKey{
+			User:     openfga.PtrString(fullUser),
+			Relation: openfga.PtrString(string(relation)),
+			Object:   openfga.PtrString(object),
+		},
+	}
+
+	data, _, err := c.apiClient.OpenFgaApi.Check(context.Background()).Body(body).Execute()
+	if err != nil {
+		logger.Debug("Failed checking permissions", logger.Ctx{"err": err})
+		return false
+	}
+
+	return data.GetAllowed()
+}
+
+// IsAdmin returns true if the provided user is a member of the "admin" group.
+func (c *Client) IsAdmin(user string) bool {
+	return c.HasPermission(user, RelationMember, GroupObject("admin"))
+}
+
+// ListObjects returns the a list of objects the user is related to. It also returns whether the user is an admin.
+func (c *Client) ListObjects(user string, relation Relation, objectType ObjectType) ([]string, error) {
+	fullUser := fmt.Sprintf("user:%s", user)
+	body := openfga.NewListObjectsRequest(string(objectType), string(relation), fullUser)
+
+	objectResponse, _, err := c.apiClient.OpenFgaApi.ListObjects(context.Background()).Body(*body).Execute()
```

Comment:
For new code we should be accepting context arguments not using context.Background.